### ISSUE TITLE
Certain knots cause knot_floer_homology to crash

### ISSUE THREAD

**Issue by agreatnate** [3-manifolds/knot_floer_homology#10]
Various knots (originally created in SageMath and converted to snappy) seem to cause the knot_floer_homology function to crash, killing the python process.  

For example, the 7-crossing knot described by the PD code: [(0, 11, 1, 12), (1, 8, 2, 9), (2, 5, 3, 6), (6, 3, 7, 4), (4, 7, 5, 8), (12, 9, 13, 10), (10, 13, 11, 0)], which appears to describe a valid knot.

Running 
```
knot_floer_homology.pd_to_hfk([(0, 11, 1, 12), (1, 8, 2, 9), (2, 5, 3, 6), (6, 3, 7, 4), (4, 7, 5, 8), (12, 9, 13, 10), (10, 13, 11, 0)])
```

Will sometimes result in a 
`ValueError: Could not compute a small girth Morse code`
But more typically (or when rerunning the same code again) will cause the python process to crash.

Some debugging indicates that PlanarDiagram::GetSmallGirthMorseCode is returning an empty list, apparently concluding that there is some issue with the planar diagram.

**Comment by NathanDunfield** [comment]
If I have SnapPy draw this knot, it looks like it's a connected sum, with an obvious twist simplification:

![untitled](https://user-images.githubusercontent.com/3139290/151622244-f5d44233-75dd-4cea-89f8-2b949be8f79f.png)

I suspect that the code for finding a low-girth diagram is implicitly assuming that the knot is prime, or at least that there is a bug specific to non-prime diagrams.  If one simplifies to eliminate the twist, it does give the right answer:
```
sage: pd = [(0, 11, 1, 12), (1, 8, 2, 9), (2, 5, 3, 6), (6, 3, 7, 4), (4, 7, 5, 8), (12, 9, 13, 10), (10, 13, 11, 0)]                                                                                  
sage: L = snappy.Link(pd)                                                                                                                                                                              
sage: L.simplify('global'); L                                                                                                                                                                          
True
<Link: 1 comp; 6 cross>
sage: L.knot_floer_homology()                                                                                                                                                                          
{'L_space_knot': False,
 'epsilon': -1,
 'fibered': True,
 'modulus': 2,
 'nu': -1,
 'ranks': {(-2, 0): 1, (-1, 1): 2, (0, 2): 3, (1, 3): 2, (2, 4): 1},
 'seifert_genus': 2,
 'tau': -2,
 'total_rank': 9}
```

**Comment by agreatnate** [comment]
That makes sense about the code having trouble with non-prime and/or unsimplified diagrams, I'd noticed that asking snappy to simplify this knot fixed the issue for this knot, however I encounter additional knots later for which a call to simplify(), (and even simplify('global')) doesn't fix the problem and python still crashes.  The unpredictability makes it hard to reproduce and even find the offending knot.  (The example I gave was one where the code seemed to dependably fail, other knots seem to work most of the time but occasionally fail...)

It isn't so great of a problem when the library raises an exception, its the crashing that is the bigger problem.

The issue seems to be that calls to be that calls to diag.GetSmallGirthMorseCode() are unpredictable, sometimes it returns an empty list and other times it does not.

This issue then manifests in pyWrapper.cpp.  Sometimes the call to `diag.GetSmallGirthMorseCode(1)` in line 216 returns a non-empty list (thus avoiding the exception there), but then the subsequent call to `diag.GetSmallGirthMorseCode()` in line 226 *does* return an empty list, (but this is never checked for). This empty MorseCode is then passed to ComputingKnotFloer which attempts (in line 27 of KnotFloer.cpp) to access the first entry of an empty list, causing a segfault.

I think it would solve the segfaulting issue to do a check after line 226 whether M is empty, and if so raise an exception (just like happens in lines 217-219 now).  

I'd test it myself but I don't know how to compile the library, attempts to run `python3 setup.py build` result in an error about being unable to find cython_src/hfk.cpp. (Maybe that is a separate issue...)

**Comment by th-rtyf-re** [comment]
Hello, I happen to be following this repository and perhaps I can give some information about something I noticed in the original ComputeHFKv2 code that might help here. If you already know this, then I apologize for interjecting.

`GetSmallGirthMorseCode` calls C++'s random function `rand` once for each attempt at making a Morse code, on line 106. I myself have encountered erratic behavior while testing because of this, in particular when calling the function multiple times in the same program. To get consistent behavior, I added a line `srand(1);` at the very beginning of the function to set the random seed (In retrospect, `srand(0);` might have been more natural).

**Comment by NathanDunfield** [comment]
> I'd test it myself but I don't know how to compile the library, attempts to run `python3 setup.py build` result in an error about being unable to find cython_src/hfk.cpp. (Maybe that is a separate issue...)

The `hfk.cpp` should be auto-generated by Cython from hkf.pyx.  I would do `python3 -m pip install Cython` and then see if you can build it.

**Comment by culler** [comment]
Actually, it appears that the diag.GetSmallGirthMorseCode method already
retries, with the default maximum number of
retries being 1000.  The default is what is currently being used.  So all
that is needed is to add a check for whether the
Morse list is empty and raise an exception if it is.

- Marc

**Comment by agreatnate** [comment]
> The hfk.cpp should be auto-generated by Cython from hkf.pyx. I would do python3 -m pip install Cython and then see if you can build knot_floer_homology.

Thanks!

Yes the diag.GetSmallGirthMorseCode() already retries many times, but not if it encounters the issue described in this bug.  I don't quite follow the logic, but if some incorrect state is reached within that code it immediately returns an empty MorseCode object. So in fact increasing the number of retries only increases the probability that this bug will be encountered.

I realized that rather than just throwing an exception when the MorseCode M is empty, the code might as well try to use the MorseCode obtained previously in LastCheckBeforeComputation (which may not have a very short girth which could result in a "Girth number exceeds..." value error being raised, still better than a segfault).

Ultimately the correct fix will require patching the code in diag.GetSmallGirthMorseCode.

**Comment by th-rtyf-re** [comment]
I've added a pull request #12 with two changes:
- I set the random seed at the beginning of the function GetSmallGirthMorseCode. This doesn't fix any bugs within the function, but it ensures consistent behavior when calling the function several times
- I deleted a part of the function that forbids adding crossings to the last (rightmost) strand. I'm not sure why this was in the code, but it causes problems in cases like the above connected sum knot with a twist. This is because the function constructs its Morse code by selecting a first crossing, and then iteratively attaching a neighboring crossing as long as it doesn't use the last strand: if no such crossing exists, then the function exists early with an empty list.

With this last change, I got
```
>>> knot_floer_homology.pd_to_hfk([(0, 11, 1, 12), (1, 8, 2, 9), (2, 5, 3, 6), (6, 3, 7, 4), (4, 7, 5, 8), (12, 9, 13, 10), (10, 13, 11, 0)])
{'L_space_knot': False, 'epsilon': -1, 'fibered': True, 'modulus': 2, 'nu': -1, 'ranks': {(-2, 0): 1, (-1, 1): 2, (0, 2): 3, (1, 3): 2, (2, 4): 1}, 'seifert_genus': 2, 'tau': -2, 'total_rank': 9}
```

I have tested the changes a little bit, but not extensively, and in particular I still do not know why there was this rule of not attaching crossings to the last strand. But I also cannot see why removing it would cause any problems. If you know anything more about this, I will gladly be corrected!

**Comment by culler** [comment]
Maybe it would be a good idea to ask Zoltán why he did that.

- Marc

**Comment by th-rtyf-re** [comment]
> Maybe it would be a good idea to ask Zoltán why he did that.

Good idea, I have sent him an email!

**Comment by th-rtyf-re** [comment]
I got the following response from Zoltán Szabó:

> Hello Isaac,
> 
> The other part of the algorithm that computes knot Floer homology from a Morse projection would work also in the case when the last strand has a crossing. So one could modify the GetSmallGirthMorseCode code to allow for that. Of course for non-prime knots there is a much simpler way to compute knot Floer homology: just use the connected sum formula for knot Floer homology.
> 
> I hope this is helpful.
> Best,
> Zoltan

So, mathematically speaking, there is no problem with removing that part of the program. It's unclear to me if there's a computational reason to do this (perhaps it's a vestige of some code related to the fact that local minima must be in position 1), but from what I can tell removing this code can only reduce the girth of the Morse code.

**Comment by NathanDunfield** [comment]
Closed by 92e514d7ffbc7a6d63e1551b6.  Thanks everyone!
